### PR TITLE
pybind/rados: add Ioctx::cmpext()

### PR DESCRIPTION
**Pull request by zjcmszh**
Signed-off-by: Zhang Jiao <zhangjiao@cmss.chinamobile.com>

### CODE REVIEW

File: src/test/pybind/test_rados.py [hunk 1]
```diff
@@ -333,6 +333,12 @@ def test_trunc(self):
         size = self.ioctx.stat('abc')[0]
         eq(size, 2)
 
+    def test_cmpext(self):
+        self.ioctx.write('test_object', b'abcdefghi')
+        eq(0, self.ioctx.cmpext('test_object', b'abcdefghi', 0))
```
Comment (ideepika) [inline src/test/pybind/test_rados.py]: think we can check for None if the function runs successfully if we update with an assert.

File: src/pybind/rados/rados.pyx [hunk 2]
```diff
@@ -3031,6 +3032,35 @@ returned %d, but should return zero on success." % (self.name, ret))
...
+        :param key: the name of the object
+        :type key: str
+        :param cmp_buf: buffer containing bytes to be compared with object contents
+        :type cmp_buf: 
+        :param offset: object byte offset at which to start the comparison
+        : type offset: int
+        
+        :raises: :class:`TypeError`
+        :raises: :class:`Error`
+        returns: 0 -on success, negative error code on failure,
```
[the remark this answers is not in the document]
Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: when the buffer containing bytes is not exactly the same as the object content ,we need to get the value of dismatch_off by returning the value, so i think we need a return value here.

File: src/pybind/rados/rados.pyx [hunk 3]
```diff
@@ -3031,6 +3032,35 @@ returned %d, but should return zero on success." % (self.name, ret))
...
+    @requires(('key', str_type), ('cmp_buf', bytes), ('offset', int))
+    def cmpext(self, key, cmp_buf, offset=0):
+        '''
+        Compare an on-disk object range with a buffer
+        :param key: the name of the object
+        :type key: str
+        :param cmp_buf: buffer containing bytes to be compared with object contents
+        :type cmp_buf: 
+        :param offset: object byte offset at which to start the comparison
+        : type offset: int
```
[the remark this answers is not in the document]
Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: updated, please check.

File: src/pybind/rados/rados.pyx (hunk 2 above)
[the remark this answers is not in the document]
Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: please review,  very very thankful.

File: src/pybind/rados/rados.pyx [hunk 4]
```diff
@@ -3031,6 +3032,35 @@ returned %d, but should return zero on success." % (self.name, ret))
...
+        self.require_ioctx_open()
+        key = cstr(key, 'key')
+        cdef:
+             char *_key = key
+             char *_cmp_buf = cmp_buf
+             size_t _cmp_buf_len = len(cmp_buf)
+             uint64_t _offset = offset
+        with nogil:
+            ret = rados_cmpext(self.io, _key, _cmp_buf, _cmp_buf_len, _offset)
+        if -4095 < ret < 0:
```
[the remark this answers is not in the document]
Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: updated，please check.

File: src/pybind/rados/rados.pyx [hunk 5]
```diff
@@ -3031,6 +3032,35 @@ returned %d, but should return zero on success." % (self.name, ret))
...
+        :type key: str
+        :param cmp_buf: buffer containing bytes to be compared with object contents
+        :type cmp_buf: bytes 
+        :param offset: object byte offset at which to start the comparison
+        :type offset: int
+        
+        :raises: :class:`TypeError`
+        :raises: :class:`Error`
+        :returns: 0 -on success, negative error code on failure,
+                 (-MAX_ERRNO - dismatch_off) on mismatch
```
[the remark this answers is not in the document]
Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: updated, please check.

**Comment by zjcmszh** [comment]
test code:
```
from rados import (Rados, WriteOpCtx, ObjectNotFound)
from nose.tools import eq_ as eq, ok_ as ok, assert_raises
import threading
import rados
import unittest

class TestIoctx(unittest.TestCase):
    def setUp(self):
        self.rados = Rados(conffile='./ceph.conf')
        self.rados.connect()
        self.rados.create_pool('pool12')
        assert self.rados.pool_exists('pool12')
        self.ioctx = self.rados.open_ioctx('pool12')
    def tearDown(self):
        self.ioctx.close()
        self.rados.delete_pool('pool12')
        self.rados.shutdown()

    def test_cmpext(self):
        self.ioctx.write('test_object', b'abcdefghi')
        eq(0, self.ioctx.cmpext('test_object', b'abcdefghi', 0))
        # MAX_ERRNO = 4095
        eq(-4095 - 4, self.ioctx.cmpext('test_object', b'abcdxxxxx', 0))

if __name__ == '__main__':
    unittest.main()

```
test results:
```
[root@jiaowa build]# python test_cmpext.py
.
----------------------------------------------------------------------
Ran 1 test in 2.905s

OK
```
@ideepika   please check, thanks a lot

File: src/pybind/rados/rados.pyx [hunk 6]
```diff
@@ -3031,6 +3032,34 @@ returned %d, but should return zero on success." % (self.name, ret))
...
+        self.require_ioctx_open()
+        key = cstr(key, 'key')
+        cdef:
+             char *_key = key
+             char *_cmp_buf = cmp_buf
+             size_t _cmp_buf_len = len(cmp_buf)
+             uint64_t _offset = offset
+        with nogil:
+            ret = rados_cmpext(self.io, _key, _cmp_buf, _cmp_buf_len, _offset)
+        assert ret < -4095 or ret == 0, "Ioctx.cmpext(%s): failed to compare %s" % (self.name, key)
```
Comment (ideepika) [inline src/pybind/rados/rados.pyx]: we discourage the use of any hardcoded values, instead, would encourage you to add `MAX_ERRNO` to C_API and make use of it, as the consumer of this API will be requiring to use.

Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: i will fix it tonight, thanks.

**Comment by zjcmszh** [comment]
test code：
```
from rados import (Rados, WriteOpCtx, ObjectNotFound, MAX_ERRNO)
from nose.tools import eq_ as eq, ok_ as ok, assert_raises
import threading
import rados
import unittest

class TestIoctx(unittest.TestCase):
    def setUp(self):
        self.rados = Rados(conffile='./ceph.conf')
        self.rados.connect()
        self.rados.create_pool('pool12')
        assert self.rados.pool_exists('pool12')
        self.ioctx = self.rados.open_ioctx('pool12')
    def tearDown(self):
        self.ioctx.close()
        self.rados.delete_pool('pool12')
        self.rados.shutdown()

    def test_cmpext(self):
        self.ioctx.write('test_object', b'abcdefghi')
        eq(0, self.ioctx.cmpext('test_object', b'abcdefghi', 0))
        eq(-MAX_ERRNO - 4, self.ioctx.cmpext('test_object', b'abcdxxxxx', 0))

if __name__ == '__main__':
    unittest.main()

```
test result:

```
[root@centos7jiaowa build]# python test_cmpext.py
.
----------------------------------------------------------------------
Ran 1 test in 2.836s

OK

```
@ideepika  updated,  please check.

File: src/pybind/rados/rados.pyx (hunk 6 above)
Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: updated, please check , thanks.

**Comment by zjcmszh** [comment]
@ideepika   i am very appriciate if you could merge these commits...

File: src/pybind/rados/rados.pyx [hunk 7]
```diff
@@ -3031,6 +3038,34 @@ returned %d, but should return zero on success." % (self.name, ret))
...
+        self.require_ioctx_open()
+        key = cstr(key, 'key')
+        cdef:
+             char *_key = key
+             char *_cmp_buf = cmp_buf
+             size_t _cmp_buf_len = len(cmp_buf)
+             uint64_t _offset = offset
+        with nogil:
+            ret = rados_cmpext(self.io, _key, _cmp_buf, _cmp_buf_len, _offset)
+        assert ret < -MAX_ERRNO or ret == 0, "Ioctx.cmpext(%s): failed to compare %s" % (self.name, key)
```
Comment (ideepika) [inline src/pybind/rados/rados.pyx]: can you please explain if ret is 0 then, why are we raising an error, since a 0 return value for success?

Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: if ret < -MAX_ERRNO or ret == 0，It doesn't throw an exception...

**Comment by zjcmszh** [comment]
@ideepika  If there are no problems, would you approve it?  vary vary vary very thankful.

File: src/pybind/rados/rados.pyx (hunk 7 above)
Comment (ideepika) [reply src/pybind/rados/rados.pyx]: thanks for explaining, lgtm!

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-12-26_05:49:41-rados-wip-kefu-testing-2019-12-26-1055-distro-basic-smithi/

failures tracked / fixed by

- https://tracker.ceph.com/issues/38539
- https://tracker.ceph.com/issues/43396
- https://github.com/ceph/ceph/pull/32427
- https://github.com/ceph/ceph/pull/32426
- https://github.com/ceph/ceph/pull/32428
- #32429

**Comment by tchaikov** [comment]
@zjcmszh needs rebase

**Comment by tchaikov** [comment]
@ideepika good to merge?

**Comment by ideepika** [comment]
jenkins test make check please

**Comment by ideepika** [comment]
> @ideepika good to merge?



> @ideepika good to merge?

looks good to me! +1

**Pull request merged**